php/libraries/GenericDownloader.class.inc

```diff
@@ -0,0 +1,155 @@
+ * suitable only when a file can be directly and simply retrieved from the
+ * file system.
+ *
+ *  @category Main
+ *  @package  Behavioural
+ *  @author   John Saigle <john.saigle@mcgill.ca>
+ *  @license  http://www.gnu.org/licenses/gpl-3.0.txt GPLv3
+ *  @link     https://www.github.com/aces/Loris
+ */
+class GenericDownloader implements FileDownloader
+{
+
+    protected $realBasePath;
+
+    /**
+     * Constructor for GenericDownloader.
+     *
+     * @param string $basePath The full path to the directory that will serve
+     *                              the files for download.
+     *
+     * @return none
+     */
+    function __construct(string $basePath)
+    {
+        $this->realBasePath = realpath($basePath);
+        // Ensure there is a trailing forward slash.
+        $this->realBasePath = rtrim($this->realBasePath, '/') . '/';
+    }
+
+    /**
+     * Downloads a file from the filesystem.
+     *
+     * @param string $userPath The full path supplied by the calling program.
+     *                          This path can be unsanitized (i.e. can accept
+     *                          raw request data) as sanitization is done
+     *                          automatically.
+     *
+     * @return none
+     */
+    public function downloadFile(string $userPath): void
+    {
+        // Ensure that every following test is conducted on a sanitized
+        // (non-malicious) file path.
+        $this->sanitizePath($userPath);
+        $fileName = basename($userPath);
+        $fullPath = $userPath;
+
+        // Ensure the server is configured correctly to allow the apache user
+        // to download files from the directory where they are stored.
+        if (!is_readable($this->realBasePath)) {
+            error_log(
+                "ERROR: $this->realBasePath should be readable by the "
+                . 'apache user but is not.'
+            );
+            http_response_code(500);
+            return;
+        }
+        // Make sure file is actually a file.
+        if (!is_file($fullPath)) {
+            error_log($fullPath);
+            error_log("ERROR: File requested for download is not a file.");
+            http_response_code(400);
+            return;
+        }
+
+        // Make sure the file itself is readable.
+        if (!is_readable($fullPath)) {
+            error_log("ERROR: A forbidden file was requested for download");
+            http_response_code(403);
+            return;
+        }
+
+        // Make sure the requested file exists.
+        if (!file_exists($fullPath)) {
+            error_log("ERROR: File requested for download does not exist.");
+            http_response_code(404);
+            return;
+        }
+
+        // Get mime type and do the file download.
+        $mimeType = mime_content_type($fullPath);
+        header("Content-type: $mimeType");
+        header("Content-Disposition: attachment; filename=$fileName");
```

Comment:
There are the responsibility of the PSR7 handler. Maybe the interface should have a `GetContentType(string $filename) : string` (or similar) method to make it possible for a real middleware to inject the header.